### PR TITLE
count number holes in parser [WIP]

### PR DESCRIPTION
**Pull request by 304NotModified**
Not sure about the impl.

It could be added to the PartList, but then we should drop inheritance IMO

### CODE REVIEW

File: src/Parser/TemplateParser.cs [hunk 1]
```diff
@@ -14,17 +14,17 @@ public static PartList Parse(string template)
...
         private readonly string _template;
         private readonly int _length;
         private int _position;
         private int _holeIndex;
-        private PartList _parts;
+        private readonly PartList _parts;
 
         private TemplateParser(string template)
         {
+            _parts = new PartList();
```
Comment (jods4) [inline src/Parser/TemplateParser.cs:27]: Right, could even be on the declaration line.

File: src/Parser/Parts/PartList.cs [hunk 2]
```diff
@@ -10,6 +10,23 @@ public PartList()
         {
         }
 
+ 
+        
+        /// <summary>
+        /// Number of <see cref="TextPart"/> with type <see cref="HoleType.Numeric"/>
+        /// </summary>
+        public int NumericPartCount { get; set; }
```
Comment (jods4) [inline src/Parser/Parts/PartList.cs:18]: What do we need those counts for? I think we only need to know if there is at least one non-numeric part, right? That could be just a bool flag set when parsing.

Comment (304NotModified) [inline src/Parser/Parts/PartList.cs:18]: yes, that is true I guess. I checked the Serilog source code, and it was testing all numeric and minimal 1 numeric AFAIK. Had that in mind...

Comment (jods4) [inline src/Parser/Parts/PartList.cs:18]: Ah, is the case when there is no hole at all important? Could we just say it's positional?

Comment (304NotModified) [inline src/Parser/Parts/PartList.cs:18]: It is for the error in Serilog I see

https://github.com/messagetemplates/messagetemplates-csharp/blob/1a49da630a1f9c9144afc280754bf2282a9eaeb0/src/MessageTemplates/MessageTemplate.cs#L60-L78

File: src/Parser/TemplateParser.cs (hunk 1 above)
Comment (304NotModified) [inline src/Parser/TemplateParser.cs:27]: I think it's good practice that we don't get nullrefs in the code, so therefor it's in the ctor and not dependend of a method.

Too bad it's worse for the perf, but I guess creating a parser and don't used it, is bad

File: src/Parser/Parts/PartList.cs (hunk 2 above)
Comment (jods4) [inline src/Parser/Parts/PartList.cs:18]: Not sure about that code quote. From the doc comment it seems this is an API to manually construct a template from tokens, and it validates that your tokens are either all positional or none.

Makes sense in that context.

I did settle for a `IsPositional` flag on the list after parsing. So no consistency issue. The flag is trivially true for a template with no hole.

File: src/Parser/TemplateParser.cs (hunk 1 above)
Comment (jods4) [inline src/Parser/TemplateParser.cs:27]: Yes, I agree. It was remanent of an old code structure I had before. Given that the ctor is private and called from inside the `Parse()` method, there's no chance you allocate an unused object ;)

I was saying that I would even move it to 
`private PartList _parts = new PartList();`
as it's not dependent on ctor params.

**Comment by 304NotModified** [comment]
Thanks for the comments & fix @jods4 !

**Pull request closed without merging**